Issue 144 rh from csv

shukon opened the pull request:
Enables creation of reports from smac2-output. Call CLI with `--file_format smac2`. There are multiple issues, including:
- we need to specify `instance_file` and `test_instance_file` in scenario-file (even if they point to the same file), because some plots require both instances... (#146)
- sometimes there are errors thrown on different parts of the analysis.
all relevant files (`run_and_results-(...).csv`, `paramstrings-(...).txt`, `param.pcs`, `traj-run-1.txt`) are expected in the specified file

AndreBiedenkapp commented on cave/plot/plotter.py:
```diff
@@ -189,6 +205,8 @@ def prepare_data(x_data):
             f.tight_layout()
             f.savefig(out)
             plt.close(f)
+        if len(output_fn) == 1:
```
This seems odd to me.
If you only have one output_fn you return a string (the first element of a list).
But if the length is != 1 then you return a list. Why is that?
I'd try to always return the same type. This is always less error prone

shukon replied:
True. I left it as it was because I plan to rework scatter and cdf anyway. This was more of a hotfix to reenable the report with only train-instances.

AndreBiedenkapp replied:
I see. For now it's ok but we shouldn't lose track of it, as it will make future maintenance more difficult

AndreBiedenkapp commented on cave/reader/smac2_reader.py:
```diff
@@ -0,0 +1,195 @@
+            runhistory and (if available) validated runhistory
+        """
+
+        validated_rh = False
```
This stays always false. We should also parse Validated SMAC2 output.

shukon replied:
I don't know about the format. If you have any validated smac2-data, I'd be glad to work on it.

AndreBiedenkapp replied:
I'll search for some on Monday.

AndreBiedenkapp commented on cave/reader/smac2_reader.py:
```diff
@@ -0,0 +1,195 @@
+                                    "\'paramstrings\'-file could be found "
+                                    "in %s" % self.folder)
+        configs_fn = os.path.join(self.folder, configs_fn.group())
+        self.logger.debug("Configurations loaded as kind-of csv from %s", configs_fn)
```
I don't know what you mean with kindo-of csv.

AndreBiedenkapp commented on cave/reader/smac2_reader.py:
```diff
@@ -0,0 +1,195 @@
+        return (rh, validated_rh)
+
+    def get_trajectory(self, cs):
+        traj_fn = os.path.join(self.folder, '../traj-run-1.txt')
```
Couldn't traj-run-1 also be called traj-run-5 or any number?
I think you should glob for traj-run-* in self.folder

shukon replied:
Right.

AndreBiedenkapp commented on cave/reader/smac2_reader.py:
```diff
@@ -0,0 +1,195 @@
+        csv_data.apply(add_to_traj, axis=1)
+        return traj
+
+    def _remove_inactive(self, cs, configuration, values):
```
Why is this necessary? Couldn't this be achived with the ConfigSpace package?

shukon replied:
Yes, I realized that later and forgot to change it.

AndreBiedenkapp commented on cave/utils/convert_for_epm.py:
```diff
@@ -78,12 +81,24 @@ def convert_data_for_epm(scenario:Scenario, runhistory:RunHistory):
+                               "is %s, if runtime then logged data is expected) "
+                               "with error message: %s", run_obj, err)
+                logger.debug(traceback.format_exc())
+            run_obj = 'runtime_'
```
Why can this fail? And why change the run_obj to 'runtime_'?
Is this an issue arising from SMAC2 input?

shukon replied:
see slack. It's something I don't really understand.

AndreBiedenkapp commented on cave/utils/csv2rh.py:
```diff
@@ -0,0 +1,250 @@
+            self.data = data
+
+        # Expecting header as described in docstring
+        valid_values = ['seed', 'cost', 'time', 'status', 'config_id', 'instance_id']
```
To make this work, in line 81 you would need to compare column_name.lower to valid_values.

AndreBiedenkapp commented on cave/utils/csv2rh.py:
```diff
@@ -0,0 +1,250 @@
+        for column_name in self.data.columns:
+            if not (column_name in valid_values or
+                    column_name.startswith('p_') or  # parameter
+                    column_name.startswith('i_')):  # instance feature
```
This is a good start. However marking parameter and instance-features via p_ and i_ is very limiting. I'd suggest to use an input flag that says how many parameter values there are. Then the first X columns that are not in "valid_values" are parameters and all other values are instance features. If the number of parameters was not specified, we can treat all columns that are not in "valid_values" as parameters. This should make this interface more general and less error prone.

AndreBiedenkapp commented on cave/utils/csv2rh.py:
```diff
@@ -0,0 +1,250 @@
+        self.data.apply(add_to_rh, axis=1)
+        return rh
+
+    def get_cs(self):
```
This method should be part of Pyimp. This would allow users to very easily use pyimp as stand-alone if they are only interested in parameter importance (and maybe don't have instance features).

AndreBiedenkapp commented on cave/utils/csv2rh.py:
```diff
@@ -0,0 +1,250 @@
+        if self.pcs and type(self.pcs) == ConfigurationSpace:
+            cs = self.pcs
+        elif self.pcs and type(self.pcs) == str:
+            parameters = [p[2:] for p in self.data.columns if p.startswith('p_')]
```
Again this is very limiting to mark the parameters with p_

AndreBiedenkapp commented on cave/utils/csv2rh.py:
```diff
@@ -0,0 +1,250 @@
+                                 str(set(parameters).difference(set(cs.get_hyperparameter_names())))))
+        else:
+            parameters = [p for p in self.data.columns if p.startswith('p_')]
+            self.logger.warning("No parameter configuration space (pcs) provided! "
```
Instead of pushing a logger warning I'd suggest we actually push a real warning.
I'd use a RuntimeWarning. This should catch a users eye more than only logging warnings.
What do you think @shukon?

AndreBiedenkapp commented on cave/utils/csv2rh.py:
```diff
@@ -0,0 +1,250 @@
+            cs = ConfigurationSpace(seed=42)
+            for p in parameters:
+                cs.add_hyperparameter(UniformFloatHyperparameter(p[2:],
+                                      lower=minima[p] - 1, upper=maxima[p] + 1))
```
Why is +- 1 necessary?

shukon replied:
I noticed problems when lower and upper were the min/max-values. Maybe some float-related problem. Maybe unnecessary. Worked for me and I didn't see any issues, do you?

AndreBiedenkapp replied:
I'm not sure.
@mfeurer do you see any issues with this when creating a config space from scratch?

AndreBiedenkapp commented on cave/utils/csv2rh.py:
```diff
@@ -0,0 +1,250 @@
+        self.cs = cs
+
+    def _interpret_status(self, status):
+        if status.upper() in ["SAT", "UNSAT", "SUCCESS"]:
```
Are we sure that status will never contain white-space? If not we should check if status.strip().upper() is in [XYZ]

shukon replied:
Jup, done.

AndreBiedenkapp commented on cave/utils/csv2rh.py:
```diff
@@ -0,0 +1,250 @@
+                config_data = pd.DataFrame(config_data, columns=header)
+                config_data.set_index('CONFIG_ID', inplace=True)
+                config_data = config_data.apply(pd.to_numeric, errors='ignore')
+                parameters = ['p_' + p for p in config_data.columns]
```
Why do you prepend 'p_' if you always ignore it later with p[2:]?

AndreBiedenkapp commented on cave/utils/csv2rh.py:
```diff
@@ -0,0 +1,250 @@
+        parameters = [p for p in self.data.columns if p.startswith('p_')]
+        if 'config_id' in self.data.columns and len(parameters) > 0:
+            raise ValueError("Define configs either via \"p_\" or \"config\" in header, not both.")
+        elif 'config_id' in self.data.columns:
```
I'm not sure if we want to support 'config_id' as input format. See one of my later comments in the test input file.

shukon replied:
I think we have to, for SMAC2. So we might as well do it for CSV...

AndreBiedenkapp replied:
Yes you are right. I completely misinterpreted how the config_id thing works. After your explanation I see that it is a valid way of handling the runhistory data

AndreBiedenkapp commented on cave/utils/csv2rh.py:
```diff
@@ -0,0 +1,250 @@
+    def _complete_instances(self):
+        self.id_to_inst_feats = {}
+        self.inst_feats_to_id = {}
+        inst_feats = [i for i in self.data.columns if i.startswith('i_')]
```
As stated before. Marking instances with 'i_' is very limiting and error prone for users.

AndreBiedenkapp commented on cave/utils/csv2rh.py:
```diff
@@ -0,0 +1,250 @@
+                self.id_to_inst_feats = {i : tuple(feat) for i, feat in self.instance_features[1]}
+                self.inst_feats_to_id = {feat : i for i, feat in self.id_to_inst_feats}
+            else:
+                self.logger.warning("Instances defined but no instance features available.")
```
Is a warning enough here?
Shouldn't we stop with an error?
I'd argue for exiting with an error as the user specified instances but no instance features are available. This makes a lot of the analysis not functioning in cave.

shukon replied:
Ok. Might aswell.

AndreBiedenkapp commented on requirements.txt:
```diff
@@ -8,9 +8,9 @@ sphinx_rtd_theme
 mpld3
 bokeh==0.12.14
 smac>=0.8.0
+configspace>=0.4.6
```
What changed from 0.4.2 to 0.4.6 that this version increase is necessary?

shukon replied:
I don't know, but it stopped throwing errors after updating. Should I investigate?

AndreBiedenkapp commented on cave/reader/configurator_run.py:
```diff
@@ -40,75 +36,61 @@ def __init__(self, folder: str, ta_exec_dir: Union[str, None]=None):
+        elif file_format == 'SMAC2':
+            self.reader = SMAC2Reader(folder, ta_exec_dir)
+        elif file_format == 'CSV':
+            self.reader = CSVReader(folder, ta_exec_dir)
```
I don't think the CSVReader is imported in the beginning.

shukon replied:
I didn't implement it yet :D:D thanks for pointing it out, kinda forgot in the smac2-support.

AndreBiedenkapp commented on cave/plot/algorithm_footprint.py:
```diff
@@ -508,7 +508,7 @@ def get_clusters(self, features_2d):
         # use number of clusters with highest silhouette score
         best_score, best_n_clusters = -1, -1
-        min_clusters, max_clusters = 2, 12
+        min_clusters, max_clusters = 2, min(features_2d.shape[0], 12)
```
Why is this change necessary?

shukon replied:
If there are less than 12 instances, this throws an Error.

AndreBiedenkapp replied:
Ok then it makes total sense

AndreBiedenkapp commented on cave/utils/convert_for_epm.py:
```diff
@@ -1,6 +1,7 @@
 #!/bin/python3
 
 import numpy as np
+import traceback
```
Is traceback still necessary in the file?

shukon replied:
No.

AndreBiedenkapp commented on cave/utils/csv2rh.py:
```diff
@@ -0,0 +1,250 @@
+import os
```
Shouldn't this file be part of cave/reader?
SMAC2 and SMAC3 readers are located there

shukon replied:
Yes.

shukon commented on cave/reader/smac2_reader.py:
```diff
@@ -0,0 +1,195 @@
+                                    "\'paramstrings\'-file could be found "
+                                    "in %s" % self.folder)
+        configs_fn = os.path.join(self.folder, configs_fn.group())
+        self.logger.debug("Configurations loaded as kind-of csv from %s", configs_fn)
```
The paramstrings-file is a mix of dict and csv... but I shorten the comment!

shukon commented on test/test_files/utils/csv2rh/sample2.csv:
```diff
@@ -0,0 +1,6 @@
+config_id,seed,cost,i_feat1,i_feat2
+configname1,1,2,0,0
```
If a config is given as a name/id, then there is an additional file necessary from which to read the parameters for that configuration. This is the case for example in the runhistory of SMAC2.

shukon commented on cave/cave_cli.py:
```diff
@@ -194,7 +198,11 @@ def main_cli(self):
         else:
             logging.basicConfig(level=logging.DEBUG)
-        logging.getLogger("smac.scenario").setLevel(logging.INFO)
+            disable_loggers = ["smac.scenario",
```
I just tried to filter unnecessary logs so it's a better overview when debugging. I like that idea and I will implement it.

shukon commented on cave/cavefacade.py:
```diff
@@ -6,14 +6,15 @@
 import typing
 import json
 import copy
+import traceback
```
I think its good to print the traceback in debug when a folder couldn't be loaded.

shukon commented on cave/cavefacade.py:
```diff
@@ -136,10 +138,12 @@ def __init__(self, folders: typing.List[str], output: str,
             except Exception as err:
                 self.logger.warning("Folder %s could not be loaded, failed "
                                     "with error message: %s", folder, err)
+                self.logger.debug(traceback.format_exc())
```
Ah. Ok :)

shukon commented:
I addressed some issues in latest commit, I'm working on the rest.

shukon commented on cave/utils/csv2rh.py:
```diff
@@ -0,0 +1,250 @@
+                                 str(set(parameters).difference(set(cs.get_hyperparameter_names())))))
+        else:
+            parameters = [p for p in self.data.columns if p.startswith('p_')]
+            self.logger.warning("No parameter configuration space (pcs) provided! "
```
Done.

shukon commented on cave/utils/csv2rh.py:
```diff
@@ -0,0 +1,250 @@
+        self.data.apply(add_to_rh, axis=1)
+        return rh
+
+    def get_cs(self):
```
Sure thing. Do you mind if I do this in a separate PR, since it introduces new version-issues?

AndreBiedenkapp replied:
No a new pr sounds reasonable

shukon commented on cave/utils/smac2_reader.py:
```diff
@@ -0,0 +1,252 @@
+import os
```
Deleted. Thanks.

shukon commented:
@AndreBiedenkapp this turned out to be more complicated than anticipated. I dropped the all-in-one-option for now, CSV needs configurations and instance-features to be defined in extra files. Everything else was a bit of a hassle because of circling dependencies. I think this should be good for now and expendable if need arises for more specific CSV-formats. I added examples for smac2 and (toyexample) for CSV. When running smac2, there is an error with forward-selection for categorical hyperparameters with yes/no-options. I propose a new issue for that, because for one this PR is very large and kind of due, for the other it's a different topic. I added some documentation for the formats and automatic api-doc-generation.

AndreBiedenkapp commented:
@shukon could you elaborate what the issue with one CSV is?

AndreBiedenkapp commented:
I see. If you'd like we can merge this pull request for now and continue working on getting one CSV input running. I find having one CSV is more intuitive and expect users to have their data available in that format.

Just one more question. Why would having a smac2 style trajectory "break" things? As far as I can see (from a very brief glance at the code) the only check needed to be done.
We would need to check if there is a column named configuration or not. If there is a configuration column in the trajectory, read it similarly to [here](https://github.com/automl/pysmac/blob/67996aabf0d12fabe079a94dca5611458ba871bd/pysmac/utils/smac_output_readers.py#L139) and feed those configuration dictionaries to the config space. But I could be mistaken here and it might not be as easy as I imagine.

Regarding your comment:
> I'm sorry that this takes so long, I just started off the wrong way with this one I think.
Don't worry. Good work takes time and I'd rather have a reader that works great from the beginning than one that needs constant bugfixes :)

AndreBiedenkapp commented:
@shukon. Is this ready for review again?

shukon commented:
@AndreBiedenkapp yes!

AndreBiedenkapp commented on cave/reader/smac3_reader.py:
```diff
@@ -0,0 +1,60 @@
+
+    def get_trajectory(self, cs):
+        traj_fn = os.path.join(self.folder, 'traj_aclib2.json')
+        traj_old_fn = os.path.join(self.folder, 'traj_old.csv')
```
This is never used and could be removed.
